Introduce the cfb_shell sample app (#11072)
zephyrproject-rtos/zephyr: samples/display/cfb_shell/README.rst

position is incremented by a multiple of the ppt.

Command example (reel_board):

.. code-block:: console

         uart:~$ cfb get_param all
         param: height=120
         param: width=250
         param: ppt=8
         param: rows=15
         param: cols=250

**get_fonts**: print the index, height and width in pixels of the static
defined fonts presented in the system.

Command example (reel_board):

.. code-block:: console

         uart:~$ cfb get_fonts
         idx=0 height=32 width=20
         idx=1 height=24 width=15
         idx=2 height=16 width=10

**set_font**: choose the font to be used by passing the font index. Only one
font can be used at a time.

Command example (reel_board):

.. code-block:: console

         uart:~$ cfb set_font 0
         Font idx=0 height=32 widht=20 set

**invert**: invert the pixel color of the display.

Command example (reel_board):

.. code-block:: console

         uart:~$ cfb invert
         Framebuffer Inverted

**print**: pass the initial column and row positions and the text in
double quotation marks when it contains spaces. If text hits the edge
of the display the remaining characters will be displayed on the next line. The

> previous printed text will be erased.

Reviewer: ```suggestion
previous printed text will be overwritten.
```

Command example (reel_board):

.. code-block:: console

         uart:~$ cfb print 60 5 ZEPHYR

> **scroll**: pass the scroll direction, vertical or horizontal, the initial

Reviewer: Thanks for the update and examples!  I'm still a bit confused about the difference between "print" and "scroll".  Is "print" the same as "scroll vertical"?

Author: @dbkinder,
 The scroll command will "move" vertically or horizontally the text from the initial position to the display boundaries. And the print will only display the text without any movement.

Reviewer: Ah, so "scroll" inserts the text and moves any existing text on the display, and "print" will overwrite existing text?  That's worth clarifying.

Author: Nope. The scroll will move the text that you passed with the scroll command. What is already on the display will be overwritten as well.

column and row positions, and the text in double quotation marks when it
contains spaces. If text hits the edge the display the remaining characters
will be displayed in the next line. The text will scroll until it hits the
display boundary, last column for horizontal and last row for vertical

> direction.

Reviewer: ```suggestion
direction.  Existing text on the display will be moved, potentially off the display.
```

.. _cfb_shell_sample:

Character Framebuffer Shell Module Sample
#########################################

Overview
********
This is a simple shell module that exercises displays using the Character
Framebuffer subsystem.

Building and Running
********************

Build the sample app by choosing the OVERLAY_CONFIG for the target board, for
example:

.. code-block:: console

         cmake -DBOARD=reel_board -DOVERLAY_CONFIG="overlay-reel_board_ssd1673.conf"

.. zephyr-app-commands::
   :zephyr-app: samples/display/cfb_shell
   :gen-args: -DOVERLAY_CONFIG="overlay-your_board_display.conf"
   :goals: run

Shell Module Command Help
=========================

.. code-block:: console

         cfb - Character Framebuffer shell commands
         Options:
                 -h, --help  :Show command help.
         Subcommands:
                 init        :[none]

>                  get_param   :<all, heigh, width, ppt, rows, cols>

Reviewer: ``height``

.. _cfb_shell_sample:

Character Framebuffer Shell Module Sample
#########################################

Overview
********

> A simple shell module that exercises displays using the Character Framebuffer

Reviewer: ```suggestion
This is a simple shell module that exercises displays using the Character Framebuffer
```

subsystem.

Building and Running
********************

> Build the sample app by choosing the OVERLAY_CONFIG for the target board. E.g.:

Reviewer: ```suggestion
Build the sample app by choosing the OVERLAY_CONFIG for the target board, for example:
```

.. code-block:: console

         cmake -DBOARD=reel_board -DOVERLAY_CONFIG="overlay-reel_board_ssd1673.conf"

.. zephyr-app-commands::
   :zephyr-app: samples/display/cfb_shell
   :host-os: unix
   :board:
   :goals: run
   :compact:

Shell Module Command Help
=========================

> .. code-block:: console

Reviewer: While this shows the command syntax, it doesn't explain what the (sub)commands actually do.

Author: Aren't the command names self-explanatory?

Reviewer 2: I agree with @dbkinder here, anything is better than "[none]" and even for those with help, it is not really easy for someone not familiar with the feature to do anything simple.

example:

.. code-block:: console

         cmake -DBOARD=reel_board

.. zephyr-app-commands::
   :zephyr-app: samples/display/cfb_shell
   :goals: run

Shell Module Command Help
=========================

.. code-block:: console

         cfb - Character Framebuffer shell commands
         Options:
                 -h, --help  :Show command help.
         Subcommands:
                 init        :[none]
                 get_device  :[none]
                 get_param   :<all, height, width, ppt, rows, cols>
                 get_fonts   :[none]
                 set_font    :<idx>
                 invert      :[none]
                 print       :<col: pos> <row: pos> <text>
                 scroll      :<dir: (vertical|horizontal)> <col: pos> <row: pos>
                              <text>
                 clear       :[none]

**init**: should be called first to initialize the display.

**get_device**: prints the display device name.

**get_param**: get the display parameters where height, width and ppt
(pixel per tile) are in pixels and the number of rows and columns. The row
position is incremented by a multiple of the ppt.

> **get_fonts**: print the index, height and width in pixels of the static

Reviewer: how do you install a new font?

defined fonts presented in the system.

**set_font**: choose the font to be used by passing the font index. Only one
font can be used at a time.

**invert**: invert the pixel color of the display.

**print**: pass the initial column and row positions and the text in
double quotation marks when it contains spaces. If text hits the edge

> the display the remain characters will be displayed in the next line. The

Reviewer: ```suggestion
of the display the remaining characters will be displayed on the next line. The
```

previous printed text will be erased.

**scroll**: pass the scroll direction, vertical or horizontal, the initial
column and row positions, and the text in double quotation marks when it
contains spaces. If text hits the edge the display the remaining characters
will be displayed in the next line. The text will scroll until it hits the
display boundary, last column for horizontal and last row for vertical
direction.

> **clear**: clear the display screen.

Reviewer: It would be great, after these descriptions, to show a nice example of how to display a phrase on the display and manipulate it (and have pictures or describe what gets displayed).

Author: I can put command examples as @nashif suggested, but at the moment there is no API, AFAIK, to draw pictures, or manipulated what is printed like inversion.

Reviewer 3: @diegosueiro I think what @dbkinder means is to give examples with screenshots in the docs.
But on adding the screenshots in the docs I already raised the question (PR #9149) if it is allowed to add binary files into the git repo. In my opinion this is not a good idea and git-lfs is not setup for the zephyr repo (@nashif any inputs on this?)

Regarding the API for drawing you have PR #6826 if you need something more fancy than CFB. CFB is just a lightweight framework for displaying text where LVGL is a full blown GUI library and I think they complement each other nicely

Building and Running
********************

Build the sample app by choosing the OVERLAY_CONFIG for the target board, for
example:

.. code-block:: console

         cmake -DBOARD=reel_board -DOVERLAY_CONFIG="overlay-reel_board_ssd1673.conf"

.. zephyr-app-commands::
   :zephyr-app: samples/display/cfb_shell
   :gen-args: -DOVERLAY_CONFIG="overlay-your_board_display.conf"
   :goals: run

Shell Module Command Help
=========================

.. code-block:: console

         cfb - Character Framebuffer shell commands
         Options:
                 -h, --help  :Show command help.
         Subcommands:
                 init        :[none]
                 get_device  :[none]
                 get_param   :<all, height, width, ppt, rows, cols>
                 get_fonts   :[none]
                 set_font    :<idx>
                 invert      :[none]
                 print       :<col: pos> <row: pos> <text>
                 scroll      :<dir: (vertical|horizontal)> <col: pos> <row: pos>
                              <text>
                 clear       :[none]

**init**: should be called first to initialize the display.

**get_device**: prints the display device name.

> **get_param**: get the display parameters where heigh, width and ppt

Reviewer: height
